core/cli/node/client.hpp

```diff
@@ -0,0 +1,671 @@
+              " FIL)",
+              kDefaultMaxRetrievePrice.atto())
+              .c_str(),
+          AttoFil,
+          {kDefaultMaxRetrievePrice})
+      max_price;
+      CLI_OPTIONAL("data-selector",
+                   "(NOT SUPPORTED) IPLD datamodel text-path selector, or IPLD "
+                   "json selector",
+                   std::string)
+      data_selector;
+      CLI_BOOL("car", "Export to a car file instead of a regular file") car;
+      CLI_BOOL("allow-local", "allow use local imports") allow_local;
+      CLI_BOOL("car-export-merkle-proof",
+               "(requires --data-selector and --car) Export data-selector "
+               "merkle proof")
+      car_export_merkle_proof;
+
+      CLI_OPTS() {
+        Opts opts;
+        from(opts);
+        provider(opts);
+        piece_cid(opts);
+        allow_local(opts);
+        car(opts);
+        max_price(opts);
+        data_selector(opts);
+        car_export_merkle_proof(opts);
+        return opts;
+      }
+    };
+
+    CLI_RUN() {
+      const Node::Api api{argm};
+
+      const CID data_cid{cliArgv<CID>(argv, 0, "dataCid")};
+      const std::string path{cliArgv<std::string>(argv, 1, "path")};
+
+      const boost::optional<CID> piece_cid = args.piece_cid.v;
+      RetrievalOrder order{
+          .client =
+              (args.from ? *args.from
+                         : cliTry(api->WalletDefaultAddress(),
+                                  "Getting address of default wallet...")),
+          .miner = (args.provider ? *args.provider : Address{})};
+
+      fmt::print("max price is {} fil ({} attofil)\n",
+                 args.max_price->fil,
+                 args.max_price->atto());
+
+      fmt::print("retrieving {} to {} {}\n",
+                 data_cid,
+                 args.car ? "car" : "file",
+                 path);
+
+      const FileRef file_ref{.path = path, .is_car = args.car};
+
+      const bool local_found = [&]() {
+        if (args.allow_local) {
+          std::vector<Import> imports = cliTry(api->ClientListImports());
+          for (const Import &import : imports) {
+            if (import.root == data_cid) {
+              fmt::print("Root: {}\n", fmt::to_string(import.root));
+              fmt::print("Local path: {}\n", import.path);
+              return true;
+            }
+          }
+        }
+        return false;
+      }();
+
+      QueryOffer fin_offer;
+      if (not local_found) {  // no local -> make retrieval
+        if (!args.provider) {
+          std::vector<QueryOffer> offers =
+              cliTry(api->ClientFindData(data_cid, piece_cid));
+          offers.erase(std::remove_if(offers.begin(),
+                                      offers.end(),
+                                      [](const QueryOffer &offer) {
+                                        return not offer.error.empty();
+                                      }),
+                       offers.end());
+
+          fin_offer = *std::min_element(
```

Comment:
What if `offers` is empty?

Reply:
This means that the method cannot continue working further. I put a throw.